### ISSUE TITLE
Lock wallet upon session close

### ISSUE THREAD

**Issue by dergigi** [joinmarket-webui/joinmarket-webui#15]
As per discussion between @dennisreimann and @AdamISZ, an unlocked wallet should be locked again if the user closes the tab or the browser window. 

We can use the `/unlock`API endpoint, see https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1136

Related: https://github.com/JoinMarket-Org/joinmarket-clientserver/issues/1121

**Comment by dnlggr** [comment]
Big +1! This is the biggest pain point when developing right now (at least for me). 😄 

However, we probably cannot guarantee that a wallet will always be locked on tab/window close right?

Anyway, with the [new API functionality](https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1136) we should be able to simply re-unlock an already unlocked wallet. That'll solve the issue.

**Comment by GBKS** [comment]
Is this an issue developers have or "regular" users?

I could see myself having two tabs open, for example one to fill in a send form, the other to look up an account number. Could be confusing if closing the second tab would lock my wallet in the first tab. A timeout sounds preferable.

**Comment by dennisreimann** [comment]
Yeah, this should be handled unobtrusively, nothing the user has to worry about :)

**Comment by dergigi** [comment]
@dennisreimann let me know if you need help with this or if someone else should take over the issue. I think it would be great to get it into v0.0.3, because if we manage to get v0.0.3 into Umbrel, it will be the first version to get some user exposure.

**Comment by AdamISZ** [comment]
Is there still an issue? I found this one pretty confusing. After https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1136 (as mentioned above) I think you should be able to unlock *any* wallet you choose, whether an existing one is unlocked (accessible) or not, but it's not entirely clear exactly how people want it to work (and indeed it needs testing; the shutdown of a wallet shuts down its subservices which is something to check on).

> I could see myself having two tabs open, for example one to fill in a send form, the other to look up an account number.

I completely didn't think about this kind of usage, but I suppose it's mostly a client side concern how to make that work without something going wrong ... if you are managing a single auth token then for sure the backend doesn't care. The thing is always to remember the main limitations of: one-wallet-open-at-a-time, one taker/maker action at a time (which includes direct non-coinjoin payments, which are treated as part of the taker "suite" of functionality, for obscure and probably dumb reasons). I suppose there's even more to unpack there about what can and can't happen simultaneously but I'm derailing at this point :)

**Comment by dennisreimann** [comment]
@AdamISZ I got to try this with the current JoinMarket master, but I cannot seem to get it to work. If there is a locked wallet, I get a `409 Conflict` when calling `/unlock`, the message being `Wallet cannot be created/opened, it is locked.`

**Comment by AdamISZ** [comment]
Thanks, will look into it shortly. @dennisreimann

**Comment by AdamISZ** [comment]
@dennisreimann the error message is helping a bit here. You are getting the `RetryableStorageError` raise, which iirc is specifically only when either (a) you try to create a wallet file when a file with that name exists or (b) there is a `.lock` file for that wallet, because the wallet is either (b1) already opened by another script for writing or (b2) an earlier instance was force shutdown (an OS shutdown or some kind of crash of the script itself).

Only (b) applies here I believe, because you're calling `/unlock`. See:

https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/4ba817e5d236559c15b08090680b9e84085d7a5e/jmclient/jmclient/wallet_rpc.py#L721-L730

**Comment by theborakompanioni** [comment]
@AdamISZ Here are my findings regarding this:
- Example 1 :x:
  - Single wallet "wallet1" existing
    - Unlock "wallet1" from "browser1": :heavy_check_mark: Success 
    - Unlock "wallet1" from "browser2": :x: `409 Conflict` "Wallet cannot be created/opened, it is locked."
      (Sutation like @dennisreimann  described above.)
      
- Example 2 :heavy_check_mark:
  - Two wallets "wallet1" and "wallet2" existing
    - Unlock "wallet1" from "browser1": :heavy_check_mark: Success
    - Unlock "wallet2" from "browser2": :heavy_check_mark: Success
      ("wallet1" on "browser1" will be locked as expected, "wallet2" can be used on "browser2")

However..

- Example 3 :x:
  - Two wallets "wallet1" and "wallet2" existing
    - Unlock "wallet1" from "browser1": :heavy_check_mark: Success
    - Unlock "wallet2" from "browser2": :heavy_check_mark: Success
      (same as Example 2 till here)
    - Unlock "wallet1" from "browser1" again: :x: `409 Conflict` "Wallet cannot be created/opened, it is locked."
      From now on "wallet1" stays unlockable (cannot be opened anymore) in all browsers, even when tried again with "wallet2" locked. "wallet2" can be locked/unlocked normally.
      
It seems this reveals some kind of corrupt state introduced by https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1136 - but I am not sure.
Other wallets can be created and used after the steps of Example 3, but only "wallet1" cannot be interacted with anymore.
This has also been reproduced with a third wallet. So: Every wallet that gets "auto-locked" by the server because another one opened, will lead to "Wallet cannot be created/opened, it is locked.".

**Comment by AdamISZ** [comment]
I see.

So 'from browser 2' would mean essentially a new client of the `jmwalletd` backend right. So, if you call `/unlock` there, you don't have the *current* JWT token that is in-use for that wallet right? It seems logical.

I was not considering 'in parallel' clients, at all. I don't think we want to support that, because we'd have to figure out how to deal with multiple writing clients simultaneously, and that's exactly what we've never supported (it's the reason we have the `.lock` file: we don't want to ever have a confusion about events in the wallet like two different txs being initiated simultaneously).

If it's not acceptable to not be able to have multiple clients simultaneously controlling the same wallet, it might need a *lot* of work, I don't think that's a sensible way to go tbh. What we *do* currently have for the CLI only, is the ability to open wallets "read only" with wallet-tool.py, so people can check their balance in a different script from the one running the yield generator. For this client UI that isn't an issue, you can check the balance any time.

**Comment by AdamISZ** [comment]
> Other wallets can be created and used after the steps of Example 3, but only "wallet1" cannot be interacted with anymore.

Ah, but this part might be something independent, I'll check that out.

**Comment by theborakompanioni** [comment]
> If it's not acceptable to not be able to have multiple clients simultaneously controlling the same wallet,

I think it is not about that. The UI does not want to have more than one wallet unlocked. Not even in different browsers.
It should have one wallet open at a time. But it should be able to reopen the currently unlocked wallet when the session token is lost . This is as far as I understand the main issue.

tldr: Unlocking a wallet with the correct password should always work. 
(If nothing than being already locked wallet is preventing it)

It would be cool, if another wallet than the currently locked wallet can be unlocked when the session token for the currently locked one is lost. But is just "nice-to-have" and might be a different case. 

> > Other wallets can be created and used after the steps of Example 3, but only "wallet1" cannot be interacted with anymore.
> 
> Ah, but this part might be something independent, I'll check that out.

Using current master branch ([`4ba817`](https://github.com/JoinMarket-Org/joinmarket-clientserver/commit/4ba817e5d236559c15b08090680b9e84085d7a5e)).

**Comment by dennisreimann** [comment]
> So 'from browser 2' would mean essentially a new client of the jmwalletd backend right. So, if you call /unlock there, you don't have the current JWT token that is in-use for that wallet right? It seems logical.

Yes, it'd also be the same if browser 1 lost the token by closing the tab and reopening the web UI in a fresh tab. Solving that case was the original intention here.

> I was not considering 'in parallel' clients, at all. I don't think we want to support that, because we'd have to figure out how to deal with multiple writing clients simultaneously

Agreed, that's out of scope and also not necessary here.

> What we do currently have for the CLI only, is the ability to open wallets "read only" with wallet-tool.py, so people can check their balance in a different script from the one running the yield generator.

That'd be a separate feature/task and out of the scope here, nevertheless that'd be great: My use case would be running everything via Joininbox and using the web UI to get a glimpse at what's going on frequently.

**Comment by AdamISZ** [comment]
As of right now it seems this must just be a bug; when a new unlock event occurs, it should shut down the existing services of the existing wallet object, one of which is the wallet service itself, which should call `wallet.close()`, which should remove the `.lock` file. But seems this is not occurring. I'll spend some time getting a client running to find where it's going wrong.

**Comment by AdamISZ** [comment]
I believe the bug is fixed with https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1160 ; let me know if there are still problems after that one.

**Comment by AdamISZ** [comment]
Ah yes, I see. My bad, that is indeed a separate bug in functionality.

 https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1136 was *intended* to let you open the same wallet, again (exactly as you describe for Example 1). But I think I can immediately see why this isn't working.

The problem here is that the `unlockwallet` endpoint function explicitly tries to open the wallet file before beginning the steps to start the wallet service:

https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/4ba817e5d236559c15b08090680b9e84085d7a5e/jmclient/jmclient/wallet_rpc.py#L720-L737

and our `open_wallet` function (found in `jmclient.wallet_utils.py`) will raise `RetryableStorageError` if the `.lock` file is in place (indicating usage) (see `jmclient.storage.Storage._create_lock` if the details are of interest).

The previous case (Example 3) was just a dumb bug, this one needs a slightly more careful approach to make sure we get the desired behaviour in both cases, i.e. when `/unlock` is being called on a new wallet, but also when it's being called on an existing one.

I think just a logic check of whether the call is being sent for the current wallet(service) might be the simplest/best way. I will add the change to https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1160

**Comment by theborakompanioni** [comment]
> @theborakompanioni let me know if [JoinMarket-Org/joinmarket-clientserver@247b58d](https://github.com/JoinMarket-Org/joinmarket-clientserver/commit/247b58d0807b9db45fa179dd8ca1c8adae88d51d) works for you, it fixed it in my test.

@AdamISZ Sorry, mistakenly posted my answer to https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1160 instead of here :no_mouth:

**Comment by AdamISZ** [comment]
Just for those following, master on the other repo now has the merge of the above ( https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1160 ) so the lock/unlock bugs should be ironed out. The other issue raised there (see https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1162 https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1163) actually turned out to be something separate - unconfirmed coins weren't being displayed with the default config setting.

**Comment by theborakompanioni** [comment]
Looking really great so far.

My tests on [4389338](https://github.com/JoinMarket-Org/joinmarket-clientserver/commit/4389338d72fabbdc24b7ebd5dc276513659a778e) included:

### Tests with single wallet
- `/unlock` > `/lock` > `/unlock` > `/lock` :heavy_check_mark: 
- `/unlock` > `/unlock` > `/lock`  :heavy_check_mark: 
- `/unlock` > `/display` > `/direct-send` > `/display` > `/unlock` > `/display` :heavy_check_mark:  (correct balance)
- `/unlock` > `/coinjoin` > `/display` > `/unlock` > `/display` :heavy_check_mark:  (correct balance)

### Test with multiple wallets
2 Wallet (alice := Wallet 1; bob := Wallet 2)
- `alice/unlock` > `alice/lock` > `bob/unlock` > `bob/lock` :heavy_check_mark: 
- `alice/unlock` > `bob/unlock` > `bob/lock` :heavy_check_mark: 
- `alice/unlock` > `alice/coinjoin` > (cj finished) > `bob/unlock` > `bob/display` :heavy_check_mark: 
- `alice/unlock` > `alice/coinjoin` > (cj NOT finished) > `bob/unlock` > `bob/display` :x: (can be reproduced consistently)
  - `all makers have sent their signatures`, `sendrawtransaction` executed, bob has the correct balance
  - but after 3 minutes`[INFO]  Timed out: transaction with txid: 49c...70 not broadcast.`, retrying the `sendrawtransaction` command
  - from now on `/session` will return `"coinjoin_in_progress": true` forever.
  - NOTE: I have never unlocked alice again till now
    - When I do `alice/unlock` > `alice/coinjoin` response is `401 Unauthorized` `{"message": "Service already started."}`
  - never saw `Coinjoin completed correctly` nor `Confirmed callback in taker, confs: 1` in the logs

I have yet to make more tests running the marker service.
Maybe I can automate these tests somehow in the coming days.

There are some thing I noticed, which I'll try to reproduce and describe better soon (just to provide a quick preview):
- At some point I was not able to stop the yieldgenerator anymore  - at least `/session` always returned `"maker_running": true` (similar to above when`coinjoin_in_progress` keeps staying `true`); no output in the logs
- Somehow I managed that jmwalletd just exited, but cleanly as supervisord reported (`INFO exited: jmwalletd (exit status 0; expected)`)

**Comment by AdamISZ** [comment]
Thanks for the test results again @theborakompanioni this is proving very useful.

To analyze these cases, I need a bit of extra context. First, as I recall you're using regtest (good!), but I also need to know the situation w.r.t. makers. Are you running the `ygrunner` script, for example, or are you using Bob here as a single maker (I'm guessing that's what you mean), so you would be starting Bob as a maker from within the web client? So, Bob's wallet is the maker wallet? And is it the only maker wallet? And how is it started.

If I'm right, in guessing that you started Bob as maker from web client, then, looking at the third and fourth bullet points:

* alice/unlock > alice/coinjoin > (cj finished) > bob/unlock > bob/display :heavy_check_mark:

This third bullet point, does it mean you had earlier started Bob's maker in web client (did you lock it after?), and then after unlocking Alice, doing a coinjoin from Alice with Bob as maker, you then unlocked Bob's wallet in the web client and showed the wallet?

For the fourth bullet point, the error condition:

* alice/unlock > alice/coinjoin > (cj NOT finished) > bob/unlock > bob/display :x: (can be reproduced consistently)

is it the same as above?

I realise that a large number of these questions may end up being irrelevant, sorry about that :)

**Comment by theborakompanioni** [comment]
> To analyze these cases, I need a bit of extra context. First, as I recall you're using regtest (good!), but I also need to know the situation w.r.t. makers. Are you running the `ygrunner` script, for example, or are you using Bob here as a single maker (I'm guessing that's what you mean), so you would be starting Bob as a maker from within the web client? So, Bob's wallet is the maker wallet? And is it the only maker wallet? And how is it started.

No - I have two joinmarket server running. I am always using our docker-compose setup (you can find it [here](https://github.com/joinmarket-webui/joinmarket-webui/tree/master/docker/regtest)). One container runs the maker service indefinitely, I do all my testing on the other instance. So "alice" and "bob" are just wallets. Stupid naming in hindsight.. will revert to "wallet1" and "wallet2" again : D

> 
> If I'm right, in guessing that you started Bob as maker from web client, then, looking at the third and fourth bullet points:
> 
>     * alice/unlock > alice/coinjoin > (cj finished) > bob/unlock > bob/display heavy_check_mark
> 
> 
> This third bullet point, does it mean you had earlier started Bob's maker in web client (did you lock it after?), and then after unlocking Alice, doing a coinjoin from Alice with Bob as maker, you then unlocked Bob's wallet in the web client and showed the wallet?

No, see answer above. This is something I also tried but did not mention as it had the expected result: It was not possible as `/coinjoin` from "alice" would be rejected with "Service is already running".

> 
> For the fourth bullet point, the error condition:
> 
>     * alice/unlock > alice/coinjoin > (cj NOT finished) > bob/unlock > bob/display x (can be reproduced consistently)
> 
> 
> is it the same as above?
No, see answer above. There is no maker service started on this instance.

I hope I have explained it in a way that is comprehensible.

> I realise that a large number of these questions may end up being irrelevant, sorry about that :)

Don't worry too much. I am very happy if I can only help in some way. Thank you for your time. It is very much appreciated.

**Comment by AdamISZ** [comment]
Thanks. That's enough for me to go on I expect. Will look into it shortly. Looks like the coinjoin state needs some cleanup.

**Comment by AdamISZ** [comment]
( *By the way, I think we drifted off the original topic of this Issue (which was fixing up lock/unlock, I think that part is done). We're discussing bugfixes to coinjoin workflow now, mainly.* )

First I reproduced the basic error condition marked by :x: above on regtest using the web client.

From one point of view, this was a pretty simple bug: it's just that `JMWalletDaemon.coinjoin_state` was not updated to `CJ_NOT_RUNNING` when the service was stopped (calling `JMWalletDaemon.stopService`) - which is part of what happens when you call `/unlock` on a new wallet (i.e. Bob/wallet2 here). So it thinks the active wallet (now wallet2) is still in taker mode.

See https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1164 for a fix.

I've tested that on regtest today in two scenarios: one browser window, alice unlock, start coinjoin, lock, bob unlock, display. Or: two browser windows, alice unlock and start coinjoin in one, bob unlock and display in other. Seems to fix the problem in both situations. Note: the coinjoin gets aborted.

So, probably worth reiterating a previous discussion point:

My statement in several places that *this only works with one wallet at a time*, is not a design choice, it is more recognition that doing multiple wallets in parallel would require significantly more work and may well violate pre-existing unspoken assumptions in the core of Joinmarket's code. Joinmarket was never written, originally, with the idea of operating more than one wallet simultaneously, in the same Python executable.

So for this reason our only sensible approach is the slightly limiting model where: if you switch wallets during some non-immediate task, including a coinjoin initiated as taker, then that task has to be cut short/cancelled. We do indeed want at least the maker to be able to keep running after we lock, but see the `TODO` comment here: 

https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1164/files#diff-2d8ba6f62a8f19a84fbf3d3abb9ded8a77bd6f09d76158b73430547f482a930bR680-R685

**Comment by AdamISZ** [comment]
I think I also managed to replicate all of your 'Tests with single wallet' successfully on https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1164/commits/d493343858fd5b1256bd30455e6334dc1e6a3053 @theborakompanioni

**Comment by theborakompanioni** [comment]
> ( _By the way, I think we drifted off the original topic of this Issue (which was fixing up lock/unlock, I think that part is done). We're discussing bugfixes to coinjoin workflow now, mainly._ )

Yes! I will do some further testing and then close the issue. Thank you!

> I've tested that on regtest today in two scenarios: one browser window, alice unlock, start coinjoin, lock, bob unlock, display. Or: two browser windows, alice unlock and start coinjoin in one, bob unlock and display in other. Seems to fix the problem in both situations. Note: the coinjoin gets aborted.

Will verify.

> So, probably worth reiterating a previous discussion point:
> 
> My statement in several places that _this only works with one wallet at a time_, is not a design choice, it is more recognition that doing multiple wallets in parallel would require significantly more work and may well violate pre-existing unspoken assumptions in the core of Joinmarket's code. Joinmarket was never written, originally, with the idea of operating more than one wallet simultaneously, in the same Python executable.

Got it. Thanks for the explanation  :+1: 

> So for this reason our only sensible approach is the slightly limiting model where: if you switch wallets during some non-immediate task, including a coinjoin initiated as taker, then that task has to be cut short/cancelled. We do indeed want at least the maker to be able to keep running after we lock, but see the `TODO` comment here:
> 
> https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1164/files#diff-2d8ba6f62a8f19a84fbf3d3abb9ded8a77bd6f09d76158b73430547f482a930bR680-R685

Makes total sense! I will post any test results to https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1164.

**Comment by theborakompanioni** [comment]
Works great now. Also a huge benefit while developing ; )

Keep in mind that taker operations will be aborted when a wallet is explicitly locked or another wallet is unlocked.

There is no further work to do related to this issue, except for some possible improvements when notifying the user about what is going on (e.g. coinjoin has been aborted).  Will close this issue now, as the original problems seem fixed.  Feel free to reopen if anyone experiences any issues related to this.

**Comment by dnlggr** [comment]
Awesome! Thanks everyone. That's a huge usability boost for the app! 😄

**Comment by AdamISZ** [comment]
[#1164](https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1164) was merged.
